### PR TITLE
feat(rule 1217): "Thread.run()" should not be called directly

### PR DESCRIPTION
**Pull request by khaes-kth**
This processes [1217](https://rules.sonarsource.com/java/RSPEC-1217) warnings.

### CODE REVIEW

**Comment by slarse** [comment]
Is this ready for review?

File: src/main/java/sorald/processor/ThreadRunProcessor.java [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+import sorald.annotations.ProcessorAnnotation;
+import spoon.reflect.code.CtInvocation;
+import spoon.reflect.declaration.CtClass;
+import spoon.reflect.declaration.CtMethod;
+import spoon.reflect.factory.Factory;
+
+@ProcessorAnnotation(key = 1217, description = "\"Thread.run()\" should not be called directly")
+public class ThreadRunProcessor extends SoraldAbstractProcessor<CtInvocation> {
+    @Override
+    protected boolean canRepairInternal(CtInvocation candidate) {
```
Comment (slarse) [inline src/main/java/sorald/processor/ThreadRunProcessor.java:12]:
```suggestion
    protected boolean canRepairInternal(CtInvocation<?> candidate) {
```

File: src/main/java/sorald/processor/ThreadRunProcessor.java [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    }
+
+    @Override
+    protected void repairInternal(CtInvocation element) {
+        Factory factory = element.getFactory();
+        CtClass<?> threadClass = factory.Class().get(Thread.class);
+
+        CtMethod<?> method = threadClass.getMethodsByName("start").get(0);
+
+        CtInvocation threadStartInvocation =
```
Comment (slarse) [inline src/main/java/sorald/processor/ThreadRunProcessor.java:30]:
```suggestion
        CtInvocation<?> threadStartInvocation =
```

File: docs/HANDLED_RULES.md [hunk 3]
```diff
@@ -343,6 +344,21 @@ class SelfAssignement {
 
 -----
 
+#### "Thread.run" should not be called directly ([Sonar Rule 1217](https://rules.sonarsource.com/java/type/Bug/RSPEC-1217))
+
+The purpose of the Thread.run() method is to execute code in a separate, dedicated thread. Calling this method directly doesn't make sense because it causes its code to be executed in the current thread.
+
+To get the expected behavior, call the Thread.start() method instead.
```
Comment (slarse) [inline docs/HANDLED_RULES.md:351]: This is copied verbatim from https://rules.sonarsource.com/java/type/Bug/RSPEC-1217, which violates SonarSource's copyright of said material. Write what the processor does instead.

File: docs/HANDLED_RULES.md [hunk 4]
```diff
@@ -343,6 +344,21 @@ class SelfAssignement {
...
+
+Example:
+```diff
+    Thread myThread = new Thread(runnable);
+-   myThread.run(); // Noncompliant
++   myThread.start(); // Compliant
```
Comment (slarse) [inline docs/HANDLED_RULES.md:357]: Slightly misleading: the processor does not change the `// Noncompliant` comment to `// Compliant`. I'd personally just remove those comments as they might cause confusion.

**Review by slarse — CHANGES_REQUESTED** [review]
Everything is here now, just have a couple of comments on the docs.

**Review by slarse — APPROVED** [review]
LGTM, ready to merge?

**Comment by khaes-kth** [comment]
> LGTM, ready to merge?

Yes!

**Comment by slarse** [comment]
Nicely done! Your first merged processor, right? :tada:

**Comment by khaes-kth** [comment]
> Nicely done! Your first merged processor, right?

As far as I remember :)

**Pull request merged**